controllers/topology/internal/extensions/patches/inline/json_patch_generator_test.go

```diff
@@ -0,0 +1,871 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inline
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/api"
+)
+
+func TestGenerate(t *testing.T) {
+	tests := []struct {
+		name    string
+		patch   *clusterv1.ClusterClassPatch
+		req     *api.GenerateRequest
+		want    *api.GenerateResponse
+		wantErr bool
+	}{
+		{
+			name: "Should generate JSON Patches with correct variable values",
+			patch: &clusterv1.ClusterClassPatch{
+				Name: "clusterName",
+				Definitions: []clusterv1.PatchDefinition{
+					{
+						Selector: clusterv1.PatchSelector{
+							APIVersion: "controlplane.cluster.x-k8s.io/v1beta1",
+							Kind:       "ControlPlaneTemplate",
+							MatchResources: clusterv1.PatchSelectorMatch{
+								ControlPlane: pointer.Bool(true),
+							},
+						},
+						JSONPatches: []clusterv1.JSONPatch{
+							// .value
+							{
+								Op:    "replace",
+								Path:  "/spec/value",
+								Value: &apiextensionsv1.JSON{Raw: []byte("1")},
+							},
+							// .valueFrom.variable
+							{
+								Op:   "replace",
+								Path: "/spec/valueFrom/variable",
+								ValueFrom: &clusterv1.JSONPatchValue{
+									Variable: pointer.String("variableA"),
+								},
+							},
+							// .valueFrom.template
+							{
+								Op:   "replace",
+								Path: "/spec/valueFrom/template",
+								ValueFrom: &clusterv1.JSONPatchValue{
+									Template: pointer.String(`template {{ .variableB }}`),
+								},
+							},
+							// variable from template takes precedent
```

Comment:
Not sure I understand this use case

Reply:
I rephrased it a bit:
```go
// template-specific variable takes precedent, if the same variable exists
// in the global and template-specific variables.
```

Essentially I just want to make sure that template-specific variables would overwrite the global variables, even though currently the variables are namespaced in a way that it's impossible to overwrite global variables with a template-specific variable.